File: go/pkg/lan/store.go
package lan

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sync"
	"time"

	"github.com/vibing/zgrnet/pkg/noise"
)

// ── MemStore ────────────────────────────────────────────────────────────────

// MemStore is an in-memory Store implementation. Fast, no I/O, but data is
// lost when the process exits. Good for testing and ephemeral LANs.
type MemStore struct {
	mu      sync.RWMutex
	members map[noise.PublicKey]*Member
}

// NewMemStore creates an empty in-memory store.
func NewMemStore() *MemStore {
	return &MemStore{
		members: make(map[noise.PublicKey]*Member),
	}
}

func (s *MemStore) Add(pk noise.PublicKey) (bool, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	if _, exists := s.members[pk]; exists {
		return false, nil
	}

	s.members[pk] = &Member{
		Pubkey:    pk,
		PubkeyHex: pk.String(),
		Labels:    []string{},
		JoinedAt:  time.Now().UTC(),
	}
	return true, nil
}

func (s *MemStore) Remove(pk noise.PublicKey) (bool, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	if _, exists := s.members[pk]; !exists {
		return false, nil
	}
	delete(s.members, pk)
	return true, nil
}

func (s *MemStore) Get(pk noise.PublicKey) *Member {
	s.mu.RLock()
	defer s.mu.RUnlock()

	m, ok := s.members[pk]
	if !ok {
		return nil
	}
	cp := *m
	cp.Labels = append([]string(nil), m.Labels...)
	return &cp
}

func (s *MemStore) IsMember(pk noise.PublicKey) bool {
	s.mu.RLock()
	defer s.mu.RUnlock()
	_, ok := s.members[pk]
	return ok
}

func (s *MemStore) List() []*Member {
	s.mu.RLock()
	defer s.mu.RUnlock()

	result := make([]*Member, 0, len(s.members))
	for _, m := range s.members {
		cp := *m
		cp.Labels = append([]string(nil), m.Labels...)
		result = append(result, &cp)
	}
	return result
}

func (s *MemStore) SetLabels(pk noise.PublicKey, labels []string) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	m, ok := s.members[pk]
	if !ok {
		return fmt.Errorf("lan: pubkey %s is not a member", pk.ShortString())
	}
	m.Labels = append([]string(nil), labels...)
	return nil
}

func (s *MemStore) RemoveLabels(pk noise.PublicKey, toRemove []string) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	m, ok := s.members[pk]
	if !ok {
		return fmt.Errorf("lan: pubkey %s is not a member", pk.ShortString())
	}

	removeSet := make(map[string]bool, len(toRemove))
	for _, l := range toRemove {
		removeSet[l] = true
	}

	filtered := make([]string, 0, len(m.Labels))
	for _, l := range m.Labels {
		if !removeSet[l] {
			filtered = append(filtered, l)
		}
	}
	m.Labels = filtered
	return nil
}

func (s *MemStore) Count() int {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return len(s.members)
}

// ── FileStore ───────────────────────────────────────────────────────────────

// FileStore wraps a MemStore and persists every mutation to a JSON file.
// Atomic writes (tmp + rename) prevent corruption on crash.
type FileStore struct {
	mem  *MemStore
	path string
}

// storeFile is the JSON structure persisted to disk.
type storeFile struct {
	Members []memberJSON `json:"members"`
}

type memberJSON struct {
	Pubkey   string   `json:"pubkey"`
	Labels   []string `json:"labels"`
	JoinedAt string   `json:"joined_at"`
}

// NewFileStore creates a file-backed store. Loads existing data from
// dataDir/members.json if it exists.
func NewFileStore(dataDir string) (*FileStore, error) {
	mem := NewMemStore()
	path := filepath.Join(dataDir, "members.json")

	fs := &FileStore{mem: mem, path: path}
	if err := fs.load(); err != nil && !os.IsNotExist(err) {
		return nil, fmt.Errorf("lan: load store %s: %w", path, err)
	}
	return fs, nil
}

func (fs *FileStore) Add(pk noise.PublicKey) (bool, error) {
	added, err := fs.mem.Add(pk)
	if err != nil || !added {
		return added, err
	}
	if err := fs.save(); err != nil {
		fs.mem.Remove(pk) // rollback
		return false, err
	}
	return true, nil
}

func (fs *FileStore) Remove(pk noise.PublicKey) (bool, error) {
	m := fs.mem.Get(pk) // snapshot before remove
	removed, err := fs.mem.Remove(pk)
	if err != nil || !removed {
		return removed, err
	}
	if err := fs.save(); err != nil {
		if m != nil {
			fs.mem.Add(pk) // rollback
			fs.mem.SetLabels(pk, m.Labels)
		}
		return false, err
	}
	return true, nil
}

func (fs *FileStore) Get(pk noise.PublicKey) *Member        { return fs.mem.Get(pk) }
func (fs *FileStore) IsMember(pk noise.PublicKey) bool       { return fs.mem.IsMember(pk) }
func (fs *FileStore) List() []*Member                        { return fs.mem.List() }
func (fs *FileStore) Count() int                             { return fs.mem.Count() }

func (fs *FileStore) SetLabels(pk noise.PublicKey, labels []string) error {
	old := fs.mem.Get(pk)
	if err := fs.mem.SetLabels(pk, labels); err != nil {
		return err
	}
	if err := fs.save(); err != nil {
		if old != nil {
			fs.mem.SetLabels(pk, old.Labels) // rollback
		}
		return err
	}
	return nil
}

func (fs *FileStore) RemoveLabels(pk noise.PublicKey, toRemove []string) error {
	old := fs.mem.Get(pk)
	if err := fs.mem.RemoveLabels(pk, toRemove); err != nil {
		return err
	}
	if err := fs.save(); err != nil {
		if old != nil {
			fs.mem.SetLabels(pk, old.Labels) // rollback
		}
		return err
	}
	return nil
}

func (fs *FileStore) load() error {
	data, err := os.ReadFile(fs.path)
	if err != nil {
		return err
	}

	var sf storeFile
	if err := json.Unmarshal(data, &sf); err != nil {
		return fmt.Errorf("lan: parse %s: %w", fs.path, err)
	}

	for _, mj := range sf.Members {
		pk, err := noise.KeyFromHex(mj.Pubkey)
		if err != nil {
			continue
		}
		fs.mem.Add(pk)

		joinedAt, _ := time.Parse(time.RFC3339, mj.JoinedAt)
		if joinedAt.IsZero() {
			joinedAt = time.Now().UTC()
		}

		labels := mj.Labels
		if labels == nil {
			labels = []string{}
		}
		fs.mem.SetLabels(pk, labels)

		// Set joined_at directly (needs lock).
		fs.mem.mu.Lock()
		if m, ok := fs.mem.members[pk]; ok {
			m.JoinedAt = joinedAt
		}
		fs.mem.mu.Unlock()
	}

	return nil
}

func (fs *FileStore) save() error {
	fs.mem.mu.RLock()
	sf := storeFile{
		Members: make([]memberJSON, 0, len(fs.mem.members)),
	}
	for _, m := range fs.mem.members {
		sf.Members = append(sf.Members, memberJSON{
			Pubkey:   m.PubkeyHex,
			Labels:   m.Labels,
			JoinedAt: m.JoinedAt.Format(time.RFC3339),
		})
	}
	fs.mem.mu.RUnlock()

	data, err := json.MarshalIndent(sf, "", "  ")
	if err != nil {
		return fmt.Errorf("lan: marshal: %w", err)
	}

	dir := filepath.Dir(fs.path)
	if err := os.MkdirAll(dir, 0700); err != nil {
		return fmt.Errorf("lan: mkdir %s: %w", dir, err)
	}

	tmp := fs.path + ".tmp"
	if err := os.WriteFile(tmp, data, 0600); err != nil {
		return fmt.Errorf("lan: write %s: %w", tmp, err)
	}
	if err := os.Rename(tmp, fs.path); err != nil {
		os.Remove(tmp)
		return fmt.Errorf("lan: rename %s → %s: %w", tmp, fs.path, err)
	}

	return nil
}

File: rust/src/lan/store.rs
//! LAN membership storage — trait + built-in implementations.
//!
//! [`Store`] is the abstract interface. Two implementations:
//! - [`MemStore`]: in-memory, no persistence (good for testing)
//! - [`FileStore`]: wraps MemStore + JSON file persistence

use crate::noise::Key;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::path::PathBuf;
use std::sync::RwLock;

/// A LAN member.
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Member {
    /// Hex-encoded public key.
    pub pubkey: String,
    /// Assigned labels.
    pub labels: Vec<String>,
    /// When the member joined (RFC 3339).
    pub joined_at: String,
}

/// Abstract store interface for LAN membership.
/// Implementations must be Send + Sync for multi-threaded HTTP servers.
pub trait Store: Send + Sync {
    fn add(&self, pk: Key) -> Result<bool, String>;
    fn remove(&self, pk: Key) -> Result<bool, String>;
    fn get(&self, pk: Key) -> Option<Member>;
    fn is_member(&self, pk: Key) -> bool;
    fn list(&self) -> Vec<Member>;
    fn count(&self) -> usize;
    fn set_labels(&self, pk: Key, labels: Vec<String>) -> Result<(), String>;
    fn remove_labels(&self, pk: Key, to_remove: &[String]) -> Result<(), String>;
}

// ── MemStore ────────────────────────────────────────────────────────────────

/// In-memory store. Fast, no I/O, data lost on process exit.
pub struct MemStore {
    inner: RwLock<HashMap<Key, Member>>,
}

impl Default for MemStore {
    fn default() -> Self {
        Self::new()
    }
}

impl MemStore {
    pub fn new() -> Self {
        MemStore {
            inner: RwLock::new(HashMap::new()),
        }
    }
}

impl Store for MemStore {
    fn add(&self, pk: Key) -> Result<bool, String> {
        let mut inner = self.inner.write().unwrap();
        if inner.contains_key(&pk) {
            return Ok(false);
        }
        inner.insert(
            pk,
            Member {
                pubkey: pk.to_hex(),
                labels: vec![],
                joined_at: now_rfc3339(),
            },
        );
        Ok(true)
    }

    fn remove(&self, pk: Key) -> Result<bool, String> {
        let mut inner = self.inner.write().unwrap();
        Ok(inner.remove(&pk).is_some())
    }

    fn get(&self, pk: Key) -> Option<Member> {
        self.inner.read().unwrap().get(&pk).cloned()
    }

    fn is_member(&self, pk: Key) -> bool {
        self.inner.read().unwrap().contains_key(&pk)
    }

    fn list(&self) -> Vec<Member> {
        self.inner.read().unwrap().values().cloned().collect()
    }

    fn count(&self) -> usize {
        self.inner.read().unwrap().len()
    }

    fn set_labels(&self, pk: Key, labels: Vec<String>) -> Result<(), String> {
        let mut inner = self.inner.write().unwrap();
        let m = inner
            .get_mut(&pk)
            .ok_or_else(|| format!("lan: pubkey {} is not a member", pk.short_hex()))?;
        m.labels = labels;
        Ok(())
    }

    fn remove_labels(&self, pk: Key, to_remove: &[String]) -> Result<(), String> {
        let mut inner = self.inner.write().unwrap();
        let m = inner
            .get_mut(&pk)
            .ok_or_else(|| format!("lan: pubkey {} is not a member", pk.short_hex()))?;
        m.labels.retain(|l| !to_remove.contains(l));
        Ok(())
    }
}

// ── FileStore ───────────────────────────────────────────────────────────────

/// JSON structure for persistence.
#[derive(Serialize, Deserialize)]
struct StoreFile {
    members: Vec<Member>,
}

/// File-backed store. Wraps MemStore and persists every mutation to JSON.
pub struct FileStore {
    mem: MemStore,
    path: PathBuf,
}

impl FileStore {
    /// Creates a file-backed store. Loads existing data if the file exists.
    pub fn new(data_dir: &str) -> Result<Self, String> {
        let path = PathBuf::from(data_dir).join("members.json");
        let mem = MemStore::new();

        if path.exists() {
            let data = std::fs::read_to_string(&path)
                .map_err(|e| format!("lan: read {:?}: {}", path, e))?;
            let sf: StoreFile = serde_json::from_str(&data)
                .map_err(|e| format!("lan: parse {:?}: {}", path, e))?;
            let mut inner = mem.inner.write().unwrap();
            for m in sf.members {
                if let Ok(pk) = Key::from_hex(&m.pubkey) {
                    inner.insert(pk, m);
                }
            }
        }

        Ok(FileStore { mem, path })
    }

    fn save(&self) -> Result<(), String> {
        let inner = self.mem.inner.read().unwrap();
        let sf = StoreFile {
            members: inner.values().cloned().collect(),
        };
        drop(inner);

        let data = serde_json::to_string_pretty(&sf)
            .map_err(|e| format!("lan: marshal: {}", e))?;

        if let Some(dir) = self.path.parent() {
            std::fs::create_dir_all(dir)
                .map_err(|e| format!("lan: mkdir {:?}: {}", dir, e))?;
        }

        let tmp = self.path.with_extension("json.tmp");
        std::fs::write(&tmp, &data)
            .map_err(|e| format!("lan: write {:?}: {}", tmp, e))?;
        std::fs::rename(&tmp, &self.path)
            .map_err(|e| format!("lan: rename {:?} → {:?}: {}", tmp, self.path, e))?;

        Ok(())
    }
}

impl Store for FileStore {
    fn add(&self, pk: Key) -> Result<bool, String> {
        let added = self.mem.add(pk)?;
        if added {
            if let Err(e) = self.save() {
                self.mem.remove(pk).ok();
                return Err(e);
            }
        }
        Ok(added)
    }

    fn remove(&self, pk: Key) -> Result<bool, String> {
        let snapshot = self.mem.get(pk);
        let removed = self.mem.remove(pk)?;
        if removed {
            if let Err(e) = self.save() {
                // Rollback: re-add member with original data.
                if let Some(m) = snapshot {
                    self.mem.add(pk).ok();
                    self.mem.set_labels(pk, m.labels).ok();
                }
                return Err(e);
            }
        }
        Ok(removed)
    }

    fn get(&self, pk: Key) -> Option<Member> { self.mem.get(pk) }
    fn is_member(&self, pk: Key) -> bool { self.mem.is_member(pk) }
    fn list(&self) -> Vec<Member> { self.mem.list() }
    fn count(&self) -> usize { self.mem.count() }

    fn set_labels(&self, pk: Key, labels: Vec<String>) -> Result<(), String> {
        let old = self.mem.get(pk).map(|m| m.labels);
        self.mem.set_labels(pk, labels)?;
        if let Err(e) = self.save() {
            // Rollback to old labels.
            if let Some(old_labels) = old {
                self.mem.set_labels(pk, old_labels).ok();
            }
            return Err(e);
        }
        Ok(())
    }

    fn remove_labels(&self, pk: Key, to_remove: &[String]) -> Result<(), String> {
        let old = self.mem.get(pk).map(|m| m.labels);
        self.mem.remove_labels(pk, to_remove)?;
        if let Err(e) = self.save() {
            if let Some(old_labels) = old {
                self.mem.set_labels(pk, old_labels).ok();
            }
            return Err(e);
        }
        Ok(())
    }
}

// ── Helpers ─────────────────────────────────────────────────────────────────

pub(crate) fn now_rfc3339() -> String {
    use std::time::{SystemTime, UNIX_EPOCH};
    let dur = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default();
    let secs = dur.as_secs();
    let days = secs / 86400;
    let time_of_day = secs % 86400;
    let hours = time_of_day / 3600;
    let mins = (time_of_day % 3600) / 60;
    let s = time_of_day % 60;
    let (y, m, d) = days_to_date(days);
    format!("{:04}-{:02}-{:02}T{:02}:{:02}:{:02}Z", y, m, d, hours, mins, s)
}

pub(crate) fn days_to_date(days: u64) -> (u64, u64, u64) {
    let z = days + 719468;
    let era = z / 146097;
    let doe = z - era * 146097;
    let yoe = (doe - doe / 1460 + doe / 36524 - doe / 146096) / 365;
    let y = yoe + era * 400;
    let doy = doe - (365 * yoe + yoe / 4 - yoe / 100);
    let mp = (5 * doy + 2) / 153;
    let d = doy - (153 * mp + 2) / 5 + 1;
    let m = if mp < 10 { mp + 3 } else { mp - 9 };
    let y = if m <= 2 { y + 1 } else { y };
    (y, m, d)
}

// ── Tests ───────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;
    use crate::noise::KeyPair;

    fn test_key() -> Key {
        KeyPair::generate().public
    }

    #[test]
    fn test_mem_add_and_get() {
        let store = MemStore::new();
        let pk = test_key();

        assert!(store.add(pk).unwrap());
        assert!(!store.add(pk).unwrap());

        let m = store.get(pk).unwrap();
        assert_eq!(m.pubkey, pk.to_hex());
        assert!(m.labels.is_empty());
        assert_eq!(store.count(), 1);
    }

    #[test]
    fn test_mem_remove() {
        let store = MemStore::new();
        let pk = test_key();

        store.add(pk).unwrap();
        assert!(store.remove(pk).unwrap());
        assert!(!store.remove(pk).unwrap());
        assert_eq!(store.count(), 0);
    }

    #[test]
    fn test_mem_labels() {
        let store = MemStore::new();
        let pk = test_key();

        store.add(pk).unwrap();
        store.set_labels(pk, vec!["admin".into(), "dev".into()]).unwrap();

        let m = store.get(pk).unwrap();
        assert_eq!(m.labels, vec!["admin", "dev"]);

        store.remove_labels(pk, &["admin".into()]).unwrap();
        let m = store.get(pk).unwrap();
        assert_eq!(m.labels, vec!["dev"]);
    }

    #[test]
    fn test_mem_list() {
        let store = MemStore::new();
        store.add(test_key()).unwrap();
        store.add(test_key()).unwrap();
        assert_eq!(store.list().len(), 2);
    }

    #[test]
    fn test_mem_is_member() {
        let store = MemStore::new();
        let pk = test_key();

        assert!(!store.is_member(pk));
        store.add(pk).unwrap();
        assert!(store.is_member(pk));
        store.remove(pk).unwrap();
        assert!(!store.is_member(pk));
    }

    #[test]
    fn test_file_persistence() {
        let dir = tempfile::tempdir().unwrap();
        let dir_str = dir.path().to_str().unwrap();

        let pk = test_key();

        {
            let store = FileStore::new(dir_str).unwrap();
            store.add(pk).unwrap();
            store.set_labels(pk, vec!["admin".into()]).unwrap();
        }

        {
            let store = FileStore::new(dir_str).unwrap();
            assert_eq!(store.count(), 1);
            let m = store.get(pk).unwrap();
            assert_eq!(m.labels, vec!["admin"]);
        }
    }
}
